Initial check in of MessageImpl refactoring

dettmoney opened the pull request:
Looking at a simple ping REST throughput test, there's a lot of processing around the HashMap operations in MessageImpl. Make explicit getters and setters to avoid hashCode and equals calls.

dettmoney commented:
#build #spawn.fullfat.buckets=com.ibm.ws.eba.jaxrs.2.1.feature_fat, com.ibm.ws.eba.jaxrs.feature_fat, com.ibm.ws.eba.jaxrs_fat, com.ibm.ws.jaxrs.2.0.cdi.1.2_fat, com.ibm.ws.jaxrs.2.0.client_fat, com.ibm.ws.jaxrs.2.0.ejb.cdi_fat, com.ibm.ws.jaxrs.2.0.webcontainer_fat, com.ibm.ws.jaxrs.2.0_fat, com.ibm.ws.jaxrs.2.0_legacy_fat, com.ibm.ws.jaxrs.2.1.cdi.2.0_fat, com.ibm.ws.jaxrs.2.1.client_fat, com.ibm.ws.jaxrs.2.1.ejb.cdi_fat,
com.ibm.ws.jaxrs.2.1.sse_fat, com.ibm.ws.jaxrs.2.1.webcontainer_fat, com.ibm.ws.jaxrs.2.1_fat, com.ibm.ws.jaxrs.2.1_fat_executor, com.ibm.ws.jaxrs.2.1_fat_extended, com.ibm.ws.jaxrs.2.x_fat_clientProps, com.ibm.ws.jaxrs.ejb.cdi_fat, com.ibm.ws.jaxrs_fat, com.ibm.ws.security.openidconnect.client_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs.config, com.ibm.ws.security.saml.sso_fat.jaxrs, com.ibm.ws.security.saml.sso_fat.jaxrs.config, com.ibm.ws.jaxws.2.2.webcontainer_fat, com.ibm.ws.eba.jaxws_fat, com.ibm.ws.jaxws.2.2_fat, com.ibm.ws.jaxws.ejb_fat, com.ibm.ws.jaxws.security_fat_lWAS, com.ibm.ws.jaxws.tools_fat, com.ibm.ws.jaxws.wab_fat, com.ibm.ws.jaxws.webcontainer_fat_lWAS, com.ibm.ws.jaxws.wsat_fat, com.ibm.ws.jaxws_fat, com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.clientcontainer_fat

dettmoney commented:
The failures in the PB are related to communication with the ldap server. The failing tests pass when run locally.

dettmoney commented:
I pushed a commit that addressed all of the review comments Jared has documented thus far.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/HTTPConduit.java:
```diff
@@ -1684,7 +1694,9 @@ protected void handleResponseInternal() throws IOException {
             updateResponseHeaders(inMessage);
-            inMessage.put(Message.RESPONSE_CODE, responseCode);
+            //Liberty code change start
+            ((MessageImpl) inMessage).setResponseCode(responseCode);
```
The cast is not needed here if you update inMessage a few lines up to be of type MessageImpl instead of Mesage.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -184,7 +328,218 @@ public void setInterceptorChain(InterceptorChain ic) {
     public Object getContextualProperty(String key) {
-        Object o = getOrDefault(key, NOT_FOUND);
+        //Liberty code change start
+        Object o = null;
```
This is where if the KEYS was a HashMap of keys to Integers, if it returned non null you could just update this complex logic with:
Integer index = KEYS.get(key);
if (index != null) {
    Object value = propertyValues[index];
    return value == NOT_FOUND ? null : value;
}

Otherwise right now if it is NOT_FOUND we still look the old way.  Not sure why we should do that.  If it isn't set, we should just return null right?

The other major flaw here is that we are doing == checks on the key.  So if you are going to do == check on the you would would have to flow down to the old way of doing things because non == versions of the key would be in the Map.  This all causes problems with consistency though and can't be done like that.  If you want an example, let me know.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -297,5 +652,1699 @@ void setContextualProperty(String key, Object v) {
+        return super.remove(key);
+    }
+    
+    @Override
```
This isn't needed in the sub class.  Its existence in the super class is sufficient

jhanders34 replied:
This get method is still there.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -297,5 +652,1699 @@ void setContextualProperty(String key, Object v) {
+        return (T) get(key.getName());
+    }
+
+    @Override
```
This isn't needed in the sub class. Its existence in the super class is sufficient

jhanders34 replied:
this put method is still there.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -297,5 +652,1699 @@ void setContextualProperty(String key, Object v) {
+    }
+
+    @Override
+    public Set<String> keySet() {
```
This applies to keySet, entrySet and values.  Do we want to support this calling Iterator.remove methods on the returned Collections?  If so it would be a custom implementation of the Collection returned and the Iterator would be able to track where we are with the current key, value, entry and be able to do the remove method based off of that.  With us having an array of getter/setter properties it will be easier to track what should be removed on the Collection iterator.

If we keep this current mechanism, then we should make a copy of the Set that is returned for super.keySet() like you do later one so we aren't modifying the parents returned Set object.

jhanders34 replied:
So you decided to keep the current mechanism and not support Iterator.remove on the Set / Collection returned from keySet, entrySet and values?  If we do that we probably should have the Iterator remove return UnsupportedOperationException.  This is a change in behavior from the previous implementation of this classes where the Iterator would actually work.  I don't know if anyone calls keySet, entrySet or values on the MessageImpl class.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -297,5 +652,1699 @@ void setContextualProperty(String key, Object v) {
+        return super.containsKey(key);
+    }
+    @Override
+    public void putAll(Map<? extends String, ? extends Object> m) {
```
Should make a copy of the Map and then as you set fields instead of calling the super way of putting all, you should, remove them from the copy of the Map and only send the subset to the super.putAll

jhanders34 replied:
Excellent solution.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -19,18 +19,28 @@
+import org.apache.cxf.transport.http.AbstractHTTPDestination;
+import org.apache.cxf.transport.https.CertConstraints;
 
 public class MessageImpl extends StringMapImpl implements Message {
```
I think there are additional methods that need to be implemented:  size, clear, isEmpty, containsValue, equals, hashCode, getOrDefault, forEach, replaceAll, putIfAbsent, remove(Object, Object), replace, computeIfAbsent, computerIfPresent, compute, merge.  We may want to just implement them and have them return UnsupportedOperationException if we don't think they are called by CXF.

jhanders34 replied:
So you implemented some as throwing UnsupportedOperationException. Was that intentional or you plan to implement those still?

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/attachment/AttachmentDeserializer.java:
```diff
@@ -132,7 +133,7 @@ public void initializeAttachments() throws IOException {
 
     protected void initializeRootMessage() throws IOException {
-        contentType = (String) message.get(Message.CONTENT_TYPE);
+        contentType = (String) ((MessageImpl) message).getContentType();
```
Marking resolved.  I commented later about at least one I still saw that didn't have the change comments.  I still need to review the files that are new again to look at the code change comments on this files.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/AbstractHTTPDestination.java:
```diff
@@ -299,25 +300,30 @@ protected void invokeComplete(final ServletContext context,
     }
 
-    protected void setupMessage(final Message inMessage,
+    protected void setupMessage(final Message message,
```
Change flags still missing below:


  ((MessageImpl) exchange.getInMessage()).setHttpRequest(new HttpServletRequestSnapshot(req));

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -119,72 +126,133 @@
-        keys.add(ENDPOINT_ADDRESS);
-        keys.add(INBOUND_MESSAGE);
-        KEYS = Collections.unmodifiableSet(keys);
+        Map<String, Integer> keymap = new HashMap<String, Integer>(TOTAL);
```
This can be significantly simplified:

Define your propertyNames during the declaration of the field, i.e.

propertyNames = {CONTENT_TYPE, PROTOCOL_HEADERS, ...

Then in the static initialize you can easily just do a for loop:

for (int i = 0; i < TOTAL; ++i) {
    keymap.put(propertyNames[i], i);
}

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -324,222 +395,17 @@ public void setInterceptorChain(InterceptorChain ic) {
         //Liberty code change end
 
-        o = getOrDefault(key, NOT_FOUND);
+        Object o = getOrDefault(key, NOT_FOUND);
```
Either this should call super.getOrDefault or this whole method should just call getOrDefault to use the impl that is in this class.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -844,1507 +561,649 @@ public Object get(Object key) {
+        }
+        
+        for (Object o : propertyValues) {
+            if (o.equals(value)) {
```
Should check for NOT_FOUND here to avoid the .equals call.  Also you need to handle o being null or you are going to get a NullPointerException

jhanders34 replied:
I added a new comment for the changes.  Marking this one resolved.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -844,1507 +561,649 @@ public Object get(Object key) {
+    }
+    
+    public Object getOrDefault(String key, Object d) {
+        Object v = super.getOrDefault(key, NOT_FOUND);
```
This logic should be like the normal get method in this class.  Not sure why you aren't checking the propertyNames first.  You should use the keymap as well here like in get.   The method should be flipped and the last line be super.getOrDefault(key, d);

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -844,1507 +561,649 @@ public Object get(Object key) {
+    public boolean replace(String key, Object oldValue, Object newValue) {
+        Integer index = KEYMAP.getOrDefault(key, KEY_NOT_FOUND);
+        if (index != KEY_NOT_FOUND) {
+            if (propertyValues[index] == oldValue) {
```
This should be .equals instead of == check. Need to handle null as well.

jhanders34 replied:
Added a new comment for the new changes.  Marking this one resolved.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -844,1507 +561,649 @@ public Object get(Object key) {
+    public boolean remove(Object key, Object value) {
+        Integer index = KEYMAP.getOrDefault(key, KEY_NOT_FOUND);
+        if (index != KEY_NOT_FOUND) {
+            if (propertyValues[index] == value) {
```
Again need to use .equals and need to handle null.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -1087,7 +982,11 @@ public boolean containsValue(Object value) {
+                if (value == null) {
+                    return true;
+                }
+            } else if (o != NOT_FOUND && o.equals(value)) {
```
You might want to change this a little.  If an equals method doesn't handle null it could get messy.
`if (value == null) {
    for (Object o : propertyValue) {
        if (o == null) {
           return true;
        }
    }
} else {
    for (Object o : propertyValues) {
        if (o != null && o != NOT_FOUND && value.equals(o) {
            return true;
        }
    }
}`

This prevents passing a null to o.equals and if value is null and it calls equals on the value passed in instead of the one in the propertyValues. We do not know the impl of those equals where the equals of the value passed in needs to work correctly by definition.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -1154,12 +1045,21 @@ public Object replace(String key, Object value) {
+                } else if (propertyValues[index].equals(oldValue)){
+                    propertyValues[index] = newValue;
+                    return true;
+                }
```
Again I recommend doing the == null check on oldValue instead of proeprtyValues[index] and then do oldValue.equals to reverse the equals.

jhanders34 replied:
Same with remove with check Value.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -1180,30 +1080,96 @@ public Object putIfAbsent(String key, Object value) {
+        Integer index = KEYMAP.getOrDefault(key, KEY_NOT_FOUND);
+        if (index != KEY_NOT_FOUND) {
+            Object newValue = remappingFunction.apply(key, propertyValues[index] == NOT_FOUND ? null : propertyValues[index]);
+            if (propertyValues[index] != NOT_FOUND) {
```
I think this if / else can be simplified to:

propertyValues[index] = newValue == null ? NOT_FOUND : newValue;

jhanders34 replied:
If remappingFunction == null this method should throw NPE.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -1180,30 +1080,96 @@ public Object putIfAbsent(String key, Object value) {
+                Object oldValue = propertyValues[index];
+                Object newValue = remappingFunction.apply(key, oldValue);
+                if (newValue != null) {
+                    // should we also add the newValue if null, or do the remove like javadoc suggests???
```
If newValue is null it should set it to NOT_FOUND, i.e.

proeprtyValues[index] = newValue == null ? NOT_FOUND : newValue;

jhanders34 replied:
If remappingFunction == null this method should throw NPE.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -1180,30 +1080,96 @@ public Object putIfAbsent(String key, Object value) {
+        return super.computeIfPresent(key, remappingFunction);
     }
     @Override
     public Object merge(String key, Object value, BiFunction<? super Object, ? super Object, ? extends Object> remappingFunction) {
```
If value == null or remappingFunction == null, this method needs to throw a NullPointerException.  Should check that first in the method.

The if (propertyValues[index] != NULL) should be if (propertyValues[index] != NULL || propertyValues[index] == null)

if newValue == null it should be NOT_FOUND, i.e. removed.  So similarly to above it should be

propertyValues[index] = newValue == null ? NOT_FOUND : newValue;

return propertyValues[index]  should be changed to return getFromPropertyArray(index);

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -478,27 +483,232 @@ public Object put(String key, Object value) {
+                return false;
+            }
+
+            for (int i = current+1; i < TOTAL; i++) {
```
I guess we could have done this simpler:
```
for (next = current+1; next < TOTAL; next++) {
   if (propertyesValues[next] != NOT_FOUND) {
...
```

then you don't need to set next = i or next = TOTAL because it is already set.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -478,27 +483,232 @@ public Object put(String key, Object value) {
+    
+    final class KeySet extends AbstractSet<String> {
+        public final int size() {
+            return this.size();
```
This will create an infinite loop.  clear() has the same issue.  You need to change to MessageImpl.this.size() and MessageImpl.this.clear();

jhanders34 replied:
Similar issue in the other impls as well.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -478,27 +483,232 @@ public Object put(String key, Object value) {
+            Map.Entry<?, ?> e = (Map.Entry<?, ?>) o;
+            String key = (String) e.getKey();
+            
+            if (containsKey(key)) {
```
Can you just use getOrDefault here like in other places instead of containsKey and get?  Will save doing a scan again

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -515,13 +725,12 @@ public void putAll(Map<? extends String, ? extends Object> m) {
     }
     @Override
     public Collection<Object> values() {
```
Similar to the HashMap you can just have this be a instance variable that defaults to null and then make one and set it if it is null.  You don't need to make one every time it is called.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -297,5 +429,955 @@ void setContextualProperty(String key, Object v) {
             put(key, v);
```
Not sure if this is called a lot, but for performance I think you can have this call putIfAbsent(key, v); instead of doing containsKey and put.

dettmoney commented:
#build #spawn.fullfat.buckets=com.ibm.ws.eba.jaxrs.2.1.feature_fat, com.ibm.ws.eba.jaxrs.feature_fat, com.ibm.ws.eba.jaxrs_fat, com.ibm.ws.jaxrs.2.0.cdi.1.2_fat, com.ibm.ws.jaxrs.2.0.client_fat, com.ibm.ws.jaxrs.2.0.ejb.cdi_fat, com.ibm.ws.jaxrs.2.0.webcontainer_fat, com.ibm.ws.jaxrs.2.0_fat, com.ibm.ws.jaxrs.2.0_legacy_fat, com.ibm.ws.jaxrs.2.1.cdi.2.0_fat, com.ibm.ws.jaxrs.2.1.client_fat, com.ibm.ws.jaxrs.2.1.ejb.cdi_fat,
com.ibm.ws.jaxrs.2.1.sse_fat, com.ibm.ws.jaxrs.2.1.webcontainer_fat, com.ibm.ws.jaxrs.2.1_fat, com.ibm.ws.jaxrs.2.1_fat_executor, com.ibm.ws.jaxrs.2.1_fat_extended, com.ibm.ws.jaxrs.2.x_fat_clientProps, com.ibm.ws.jaxrs.ejb.cdi_fat, com.ibm.ws.jaxrs_fat, com.ibm.ws.security.openidconnect.client_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs.config, com.ibm.ws.security.saml.sso_fat.jaxrs, com.ibm.ws.security.saml.sso_fat.jaxrs.config, com.ibm.ws.jaxws.2.2.webcontainer_fat, com.ibm.ws.eba.jaxws_fat, com.ibm.ws.jaxws.2.2_fat, com.ibm.ws.jaxws.ejb_fat, com.ibm.ws.jaxws.security_fat_lWAS, com.ibm.ws.jaxws.tools_fat, com.ibm.ws.jaxws.wab_fat, com.ibm.ws.jaxws.webcontainer_fat_lWAS, com.ibm.ws.jaxws.wsat_fat, com.ibm.ws.jaxws_fat, com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.clientcontainer_fat

dettmoney commented:
#build #spawn.fullfat.buckets=com.ibm.ws.eba.jaxrs.2.1.feature_fat, com.ibm.ws.eba.jaxrs.feature_fat, com.ibm.ws.eba.jaxrs_fat, com.ibm.ws.jaxrs.2.0.cdi.1.2_fat, com.ibm.ws.jaxrs.2.0.client_fat, com.ibm.ws.jaxrs.2.0.ejb.cdi_fat, com.ibm.ws.jaxrs.2.0.webcontainer_fat, com.ibm.ws.jaxrs.2.0_fat, com.ibm.ws.jaxrs.2.0_legacy_fat, com.ibm.ws.jaxrs.2.1.cdi.2.0_fat, com.ibm.ws.jaxrs.2.1.client_fat, com.ibm.ws.jaxrs.2.1.ejb.cdi_fat,
com.ibm.ws.jaxrs.2.1.sse_fat, com.ibm.ws.jaxrs.2.1.webcontainer_fat, com.ibm.ws.jaxrs.2.1_fat, com.ibm.ws.jaxrs.2.1_fat_executor, com.ibm.ws.jaxrs.2.1_fat_extended, com.ibm.ws.jaxrs.2.x_fat_clientProps, com.ibm.ws.jaxrs.ejb.cdi_fat, com.ibm.ws.jaxrs_fat, com.ibm.ws.security.openidconnect.client_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs.config, com.ibm.ws.security.saml.sso_fat.jaxrs, com.ibm.ws.security.saml.sso_fat.jaxrs.config, com.ibm.ws.jaxws.2.2.webcontainer_fat, com.ibm.ws.eba.jaxws_fat, com.ibm.ws.jaxws.2.2_fat, com.ibm.ws.jaxws.ejb_fat, com.ibm.ws.jaxws.security_fat_lWAS, com.ibm.ws.jaxws.tools_fat, com.ibm.ws.jaxws.wab_fat, com.ibm.ws.jaxws.webcontainer_fat_lWAS, com.ibm.ws.jaxws.wsat_fat, com.ibm.ws.jaxws_fat, com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.clientcontainer_fat

andymc12 commented on dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/AbstractHTTPDestination.java:
```diff
@@ -268,7 +267,9 @@ public void invoke(final ServletConfig config,
         copyKnownRequestAttributes(req, inMessage);
 
-        inMessage.put(HttpServletResponse.class, resp); // Liberty change - reqd for SSE see LibertySseEventSinkImpl
+        //Liberty code change start
```
no need for the comment here or on line 272, since there is already a comment on line 271.

jhanders34 replied:
I think Andy wanted you to leave the comment you just removed and to remove the new comments you added.  The one you just removed has extra details that it is good not to lose.

andymc12 commented on dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/AbstractHTTPDestination.java:
```diff
@@ -299,25 +300,32 @@ protected void invokeComplete(final ServletContext context,
 
-    protected void setupMessage(final Message inMessage,
+    //Liberty code change start
+    protected void setupMessage(final Message message,
```
I might be missing something, but I don't think these changes (this line, line 323, lines 327-328) are necessary.  IIUC, the only time that the MessageImpl is used is on line 340 which is already casted to a MessageImpl - and already surrounded by the change tags.

dettmoney replied:
inMessage is used after the DelegatingInputStream within the method.

andymc12 commented on dev/com.ibm.ws.org.apache.cxf.cxf.rt.rs.service.description.3.2/src/org/apache/cxf/jaxrs/model/wadl/WadlGenerator.java:
```diff
@@ -0,0 +1,2333 @@
+
+    protected void doFilter(ContainerRequestContext context, Message m) {
+        //Liberty code change start
+        if (!"GET".equals(((MessageImpl) m).getHttpRequestMethod())) {
```
If I read this right, we are pulling in the WadlGenerator source file and updating the bnd file just for this change.  I don't think very many customers use Wadl (it's pretty much been replaced by Swagger/OpenAPI).  Are you seeing specific performance improvements in your benchmarks with this change?  If so, I'd probably update the benchmarks to use OpenAPI instead. :) If not, I'd say let's undo this change and the bnd change - it's one less divergence from the Apache CXF repo.

dettmoney replied:
What API should I look for in my app to determine what's bringing in the WadlGenerator? I'm not explicitly using wadl (just pinging a simple REST endpoint). The only reason I changed that code is because it's showing up in the profiles as a caller into MessageImpl.get. JAXRSUtils.runContainerRequestFilters -> WadlGenerator.filter -> WadlGenerator.doFilter is the path I observe

andymc12 replied:
That's unfortunate - it looks like we always load and register that provider... In the base CXF code, they only load it if the service.description module is on the class path - and we have it on the class path, so that means that we're executing that filter on every request...

To remove it now would probably break zero-migration - so we should probably fix it to be faster if possible.  Thanks

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/test/com/ibm/ws/jaxrs/test/MessageImplTest.java:
```diff
@@ -0,0 +1,525 @@
+    @Test
+    public void testMessageImpl() throws Exception {
+        MessageImpl message = new MessageImpl();
+        String[] propertyNames = message.getPropertyNames();
```
I was hoping that you wouldn't need to have the getGetters, getSetters, getRemoves, getContains in MessageImpl.  You can build that dynamically using reflection and it will be less fragile and the test can automatically pick up new getters, etc and fail tests when new methods are added.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -293,9 +470,965 @@ public void resetContextCache() {
+            Object val = getOrDefault(key, NOT_FOUND);
+            if (val != NOT_FOUND) {
+                AbstractMap.SimpleEntry<?,?> entry = new AbstractMap.SimpleEntry<>(key, val);
+                return entry.equals(e);
```
This should be return e.equals(entry);

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -293,9 +470,965 @@ public void resetContextCache() {
+            if (next == TOTAL) {
+                return false;
+            }
+
```
This should have a conditional in it that says
```
if (next > current) {
    return true;
}
```

This handles the hasNext(), hasNext() being called back to back without calling next.  No reason to calculate next again if we already did it.

dettmoney commented:
I've pushed changes that have addressed the comments other than the WadlGenerator issue. Still need an explanation of how that's getting into the path.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/utils/HttpUtils.java:
```diff
@@ -561,24 +591,28 @@ public static String getEndpointAddress(Message m) {
+                //Liberty code change start
+                HttpServletRequest request = (HttpServletRequest) ((MessageImpl) m).getHttpRequest();
                 Object property = request != null ? request.getAttribute("org.apache.cxf.transport.endpoint.address") : null;
                 //Liberty code change start
```
The line 599 is still there.

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/bnd.bnd:
```diff
@@ -29,3 +30,7 @@ instrument.classesExcludes: org/apache/cxf/common/jaxb/*.class
   com.ibm.wsspi.org.osgi.core
+
+-testpath: \
+	../build.sharedResources/lib/junit/old/junit.jar;version=file, \
```
There should be a ", \" at the end of the testpath.

dettmoney commented:
#build #spawn.fullfat.buckets=com.ibm.ws.eba.jaxrs.2.1.feature_fat, com.ibm.ws.eba.jaxrs.feature_fat, com.ibm.ws.eba.jaxrs_fat, com.ibm.ws.jaxrs.2.0.cdi.1.2_fat, com.ibm.ws.jaxrs.2.0.client_fat, com.ibm.ws.jaxrs.2.0.ejb.cdi_fat, com.ibm.ws.jaxrs.2.0.webcontainer_fat, com.ibm.ws.jaxrs.2.0_fat, com.ibm.ws.jaxrs.2.0_legacy_fat, com.ibm.ws.jaxrs.2.1.cdi.2.0_fat, com.ibm.ws.jaxrs.2.1.client_fat, com.ibm.ws.jaxrs.2.1.ejb.cdi_fat,
com.ibm.ws.jaxrs.2.1.sse_fat, com.ibm.ws.jaxrs.2.1.webcontainer_fat, com.ibm.ws.jaxrs.2.1_fat, com.ibm.ws.jaxrs.2.1_fat_executor, com.ibm.ws.jaxrs.2.1_fat_extended, com.ibm.ws.jaxrs.2.x_fat_clientProps, com.ibm.ws.jaxrs.ejb.cdi_fat, com.ibm.ws.jaxrs_fat, com.ibm.ws.security.openidconnect.client_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs, com.ibm.ws.security.openidconnect.server_fat.jaxrs.config, com.ibm.ws.security.saml.sso_fat.jaxrs, com.ibm.ws.security.saml.sso_fat.jaxrs.config, com.ibm.ws.jaxws.2.2.webcontainer_fat, com.ibm.ws.eba.jaxws_fat, com.ibm.ws.jaxws.2.2_fat, com.ibm.ws.jaxws.ejb_fat, com.ibm.ws.jaxws.security_fat_lWAS, com.ibm.ws.jaxws.tools_fat, com.ibm.ws.jaxws.wab_fat, com.ibm.ws.jaxws.webcontainer_fat_lWAS, com.ibm.ws.jaxws.wsat_fat, com.ibm.ws.jaxws_fat, com.ibm.ws.jaxws.cdi_fat, com.ibm.ws.jaxws.clientcontainer_fat

dettmoney commented:
Addressed Jared's code review comments and put fixes in for the failing FATs

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -152,11 +152,12 @@
     private Set<String> keySet = null;
     private Set<Map.Entry<String, Object>> entrySet = null;
-    
+    private static Object[] NOT_SET_ARRAY = new Object[TOTAL];
```
This should be final

dettmoney replied:
done

jhanders34 commented on dev/com.ibm.ws.org.apache.cxf.cxf.core.3.2/src/org/apache/cxf/message/MessageImpl.java:
```diff
@@ -293,9 +431,959 @@ public void resetContextCache() {
+        }
+        return super.merge(key, value, remappingFunction);
+    }
+    public String[] getPropertyNames() {
```
getPropertyNames isn't used any longer.  I think it can be removed.

jhanders34 commented:
#run-libby-run

jhanders34 commented:
#run-libby-run

dettmoney commented:
#run-libby-run

dettmoney commented:
Changes were tested with pingperf application, gaining 2-4% throughput performance.
